neonvm/controllers/metrics.go

```diff
@@ -68,15 +68,44 @@ type wrappedReconciler struct {
 	failing map[client.ObjectKey]struct{}
 }
 
-// WithMetrics wraps a given Reconciler with metrics capabilities.
-func WithMetrics(reconciler reconcile.Reconciler, rm ReconcilerMetrics, cntrlName string) reconcile.Reconciler {
-	return &wrappedReconciler{
+// ReconcilerWithMetrics is a Reconciler produced by WithMetrics that can return a snapshot of the
+// state backing the metrics.
+type ReconcilerWithMetrics interface {
+	reconcile.Reconciler
+
+	Snapshot() ReconcileSnapshot
+}
+
+// ReconcileSnapshot provides a glimpse into the current state of ongoing reconciles
+//
+// This type is (transitively) returned by the controller's "dump state" HTTP endpoint, and exists
+// to allow us to get deeper information on the metrics - we can't expose information for every
+// VirtualMachine into the metrics (it'd be too high cardinality), but we *can* make it available
+// when requested.
+type ReconcileSnapshot struct {
+	// ControllerName is the name of the controller: virtualmachine or virtualmachinemigration.
+	ControllerName string `json:"controllerName"`
+
+	// Failing is the list of objects currently failing to reconcile
+	Failing []string `json:"failing"`
+}
+
+// WithMetrics wraps a given Reconciler with metrics capabilities, also returning a function that
+// produces a snapshot of the metrics, for easier inspection.
+func WithMetrics(
+	reconciler reconcile.Reconciler,
+	rm ReconcilerMetrics,
+	cntrlName string,
+) ReconcilerWithMetrics {
+	r := &wrappedReconciler{
```

Comment:
nit: you can `return` on this line already, diff should be more compact.

Reply:
ah right, good catch! thanks :)